arwen/contexts/async.go

```diff
@@ -0,0 +1,885 @@
+		if err != nil {
+			return err
+		}
+		shouldLockGas = true
+	}
+	if call.ErrorCallback != "" {
+		err := runtime.ValidateCallbackName(call.ErrorCallback)
+		if err != nil {
+			return err
+		}
+		shouldLockGas = true
+	}
+
+	if shouldLockGas {
+		call.GasLocked = metering.ComputeGasLockedForAsync()
+	}
+
+	err := context.addAsyncCall(groupID, call)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RegisterLegacyAsyncCall builds a legacy AsyncCall from provided arguments,
+// computes the gas to lock depending on legacy configuration (non-dynamic gas
+// locking), then adds the AsyncCall to the predefined legacy
+// call group and informs Wasmer to stop contract execution with
+// BreakpointAsyncCall (adding the AsyncCall consumes its gas entirely).
+func (context *asyncContext) RegisterLegacyAsyncCall(address []byte, data []byte, value []byte) error {
+	if !context.canRegisterLegacyAsyncCall() {
+		return arwen.ErrLegacyAsyncCallInvalid
+	}
+
+	legacyGroupID := arwen.LegacyAsyncCallGroupID
+	_, exists := context.GetCallGroup(legacyGroupID)
+	if exists {
+		return arwen.ErrOnlyOneLegacyAsyncCallAllowed
+	}
+
+	gasToLock, err := context.computeGasLockForLegacyAsyncCall()
+	if err != nil {
+		return err
+	}
+
+	metering := context.host.Metering()
+	gasLimit := math.SubUint64(metering.GasLeft(), gasToLock)
+
+	err = context.addAsyncCall(legacyGroupID, &arwen.AsyncCall{
+		Status:          arwen.AsyncCallPending,
+		Destination:     address,
+		Data:            data,
+		ValueBytes:      value,
+		SuccessCallback: arwen.CallbackFunctionName,
+		ErrorCallback:   arwen.CallbackFunctionName,
+		GasLimit:        gasLimit,
+		GasLocked:       gasToLock,
+	})
+	if err != nil {
+		return err
+	}
+
+	context.host.Runtime().SetRuntimeBreakpointValue(arwen.BreakpointAsyncCall)
+
+	return nil
+}
+
+func (context *asyncContext) canRegisterLegacyAsyncCall() bool {
+	vmInput := context.host.Runtime().GetVMInput()
+	noGroups := len(context.asyncCallGroups) == 0
+	notInCallback := vmInput.CallType != vm.AsynchronousCallBack
+
+	return noGroups && notInCallback
+}
+
+// addAsyncCall adds the provided AsyncCall to the specified AsyncCallGroup
+func (context *asyncContext) addAsyncCall(groupID string, call *arwen.AsyncCall) error {
+	runtime := context.host.Runtime()
+	metering := context.host.Metering()
+
+	// TODO discuss
+	// TODO add exception for the first callback instance of the same address,
+	// which must be allowed to modify the AsyncContext
+	scOccurrences := runtime.CountSameContractInstancesOnStack(runtime.GetSCAddress())
+	callType := runtime.GetVMInput().CallType
+	modifiableAsyncContext := (scOccurrences == 0) || (callType == vm.AsynchronousCallBack)
+	if !modifiableAsyncContext {
+		return arwen.ErrAsyncContextUnmodifiableUnlessFirstSCOrFirstCallback
+	}
+
+	err := metering.UseGasBounded(call.GasLocked)
+	if err != nil {
+		return err
+	}
+	err = metering.UseGasBounded(call.GasLimit)
```

Comment:
It is used here necause the contract is still running, and can also create new async calls afterwards. So we consume the gas completely on the current contract instance when the async call is declared, in order to prevent the contract from creating more async calls than it can afford.

Also, it's for balanced gas management: gas is consumed on the current instance, so the gas amount can be transferred to another instance (the destination contract), or sent across shards.